### PR TITLE
Add unit tests for org.wso2.lsp4intellij.utils.FileUtils

### PR DESCRIPTION
**Pull request by ThomasPerkins1123**
Hi,

As suggested I have written some tests for `org.wso2.lsp4intellij.utils.FileUtils` with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests should help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other particular classes that you consider important.

### CODE REVIEW

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    PowerMockito.when(LocalFileSystem.getInstance()).thenThrow(URISyntaxException.class);
+    Assert.assertNull(FileUtils.virtualFileFromURI("foobar"));
+  }
+
+  @Test
+  public void testVFSToURI() {
+    Assert.assertNull(FileUtils.VFSToURI((new LightVirtualFile())));
+  }
+
+  @PrepareForTest({FileUtils.class, System.class})
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: remove this

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    PowerMockito.mockStatic(System.class);
+    PowerMockito.when(System.getProperty(Mockito.anyString())).thenReturn("Linux");
+
+    Assert.assertNull(FileUtils.sanitizeURI(null));
+
+    Assert.assertEquals("fooBar", FileUtils.sanitizeURI("fooBar"));
+    Assert.assertEquals("file:///fooBar", FileUtils.sanitizeURI("file://fooBar"));
+  }
+
+  @PrepareForTest({FileUtils.class, System.class})
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: remove system.class

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+  @PrepareForTest(LocalFileSystem.class)
+  @Test
+  public void testURIToVFS() {
+    PowerMockito.mockStatic(LocalFileSystem.class);
+    PowerMockito.when(LocalFileSystem.getInstance()).thenThrow(URISyntaxException.class);
+    Assert.assertNull(FileUtils.URIToVFS("foobar"));
+  }
+
+  @PrepareForTest({FileUtils.class, Editor.class, File.class})
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: remove  Editor.class, File.class

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    Mockito.when(editor.getProject().getBasePath()).thenReturn("test");
+
+    File file = PowerMockito.mock(File.class);
+    Mockito.when(file.getAbsolutePath()).thenReturn("fooBar");
+    PowerMockito.whenNew(File.class).withAnyArguments().thenReturn(file);
+
+    Assert.assertEquals("fooBar", FileUtils.editorToProjectFolderPath(editor));
+  }
+
+  @PrepareForTest({FileUtils.class, Project.class})
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: remove project.class

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+    PowerMockito.spy(FileUtils.class);
+    Project project = PowerMockito.mock(Project.class);
+    Mockito.when(project.getBasePath()).thenReturn("test");
+    PowerMockito.when(FileUtils.pathToUri(Mockito.anyString())).thenReturn("fooBar");
+
+    Assert.assertEquals("fooBar", FileUtils.projectToUri(project));
+  }
+
+  @PrepareForTest({VirtualFile.class})
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: remove this

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    Mockito.when(virtualFile2.getUrl()).thenReturn("");
+
+    Assert.assertFalse(FileUtils.isFileSupported(null));
+    Assert.assertFalse(FileUtils.isFileSupported(new BinaryLightVirtualFile()));
+    Assert.assertFalse(FileUtils.isFileSupported(virtualFile1));
+
+    Assert.assertTrue(FileUtils.isFileSupported(virtualFile2));
+  }
+
+  @PrepareForTest({VirtualFile.class, FileDocumentManager.class})
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: remove VirtualFile.class

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+@RunWith(PowerMockRunner.class)
+public class FileUtilsTest {
+
+  @PrepareForTest({FileEditorManagerEx.class, FileUtils.class, Editor.class, File.class, FileEditorManager.class, FileEditor.class, FileEditor[].class})
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: remove this

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  @Test
+  public void testEditorFromVirtualFile() {
+    VirtualFile file = PowerMockito.mock(VirtualFile.class);
+    Project project = PowerMockito.mock(Project.class);
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:47]: Add a blank line here as these are the parameters

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    Editor editor = PowerMockito.mock(Editor.class);
+    TextEditor textEditor = PowerMockito.mock(TextEditor.class);
+    PowerMockito.when(textEditor.getEditor()).thenReturn(editor);
+
+    FileEditor[] array = new FileEditor[]{textEditor};
+    FileEditor[] arrayEmpty = new FileEditor[0];
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: These two are used once and can be implemented directly into the mock

**Review by EricHetti — CHANGES_REQUESTED** [review]
Try to be consistent with PowerMockito and Mockito as some methods you use both

File: build.gradle [hunk 10]
```diff
@@ -23,6 +23,10 @@ intellij {
 dependencies {
     compile group: 'org.eclipse.lsp4j', name: 'org.eclipse.lsp4j', version: '0.6.0'
     compile group: 'com.vladsch.flexmark', name: 'flexmark', version: '0.34.58'
+    testCompile group: 'junit', name: 'junit', version: '4.11'
+    testCompile group: 'org.mockito', name: 'mockito-all', version: '1.6.5'
+    testCompile group: 'org.powermock', name: 'powermock-api-mockito', version: '1.6.5'
+    testCompile group: 'org.powermock', name: 'powermock-module-junit4', version: '1.6.5'
```
Comment (Braavos6) [inline build.gradle:29]: I think the versions are clashing.. I am not sure how it works.. This is the wiki for mocking https://github.com/powermock/powermock/wiki/mockito

Mainly stating that powermock has to be a lower version than mockito

Comment (Braavos6) [reply build.gradle:29]: If you are adding both maybe only use Powermock?

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java (hunk 7 above)
Comment (ThomasPerkins1123) [reply src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: Most of these could be removed but there was one that was still needed

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java (hunk 8 above)
Comment (ThomasPerkins1123) [reply src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:47]: Not quite sure what you mean here?

File: build.gradle [hunk 11]
```diff
@@ -23,6 +23,10 @@ intellij {
 dependencies {
     compile group: 'org.eclipse.lsp4j', name: 'org.eclipse.lsp4j', version: '0.6.0'
     compile group: 'com.vladsch.flexmark', name: 'flexmark', version: '0.34.58'
+    testCompile group: 'junit', name: 'junit', version: '4.11'
+    testCompile group: 'org.mockito', name: 'mockito-all', version: '1.6.5'
```
Comment (JohnLBergqvist) [inline build.gradle]: Use version `1.10.19` as that is compatible with the version of powermock you are using.

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+    Assert.assertNull(FileUtils.editorFromVirtualFile(file, project));
+  }
+
+  @PrepareForTest(LocalFileSystem.class)
+  @Test
+  public void testVirtualFileFromURI() {
+    PowerMockito.mockStatic(LocalFileSystem.class);
+    PowerMockito.when(LocalFileSystem.getInstance()).thenThrow(URISyntaxException.class);
+    Assert.assertNull(FileUtils.virtualFileFromURI("foobar"));
```
Comment (JohnLBergqvist) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:84]: Can you not test a non-null case here?

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  @Test
+  public void testVFSToURI() {
+    Assert.assertNull(FileUtils.VFSToURI((new LightVirtualFile())));
```
Comment (JohnLBergqvist) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:97]: Same as above.

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+    Assert.assertEquals("file:///Foo:/Bar", FileUtils.sanitizeURI("file:foo%3A/Bar"));
+  }
+
+  @PrepareForTest(LocalFileSystem.class)
+  @Test
+  public void testURIToVFS() {
+    PowerMockito.mockStatic(LocalFileSystem.class);
+    PowerMockito.when(LocalFileSystem.getInstance()).thenThrow(URISyntaxException.class);
+    Assert.assertNull(FileUtils.URIToVFS("foobar"));
```
Comment (JohnLBergqvist) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:138]: Same as above

**Review by JohnLBergqvist — CHANGES_REQUESTED** [review]

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+  public void testEditorFromVirtualFile() {
+    VirtualFile file = PowerMockito.mock(VirtualFile.class);
+    Project project = PowerMockito.mock(Project.class);
+    
+    Editor editor = PowerMockito.mock(Editor.class);
+    TextEditor textEditor = PowerMockito.mock(TextEditor.class);
+    PowerMockito.when(textEditor.getEditor()).thenReturn(editor);
+
+    FileEditorManagerEx fileEditorManagerEx = PowerMockito.mock(FileEditorManagerEx.class);
+
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: Remove blank line.. try to keep the mocking for the object defined in one block.. its easier to read :)

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+    Assert.assertNull(FileUtils.editorFromVirtualFile(file, project));
+  }
+
+  @PrepareForTest(LocalFileSystem.class)
+  @Test
+  public void testVirtualFileFromURI() {
+    LocalFileSystem localFileSystem = PowerMockito.mock(LocalFileSystem.class);
+    PowerMockito.mockStatic(LocalFileSystem.class);
+    PowerMockito.when(LocalFileSystem.getInstance()).thenReturn(localFileSystem);
+    PowerMockito.when(LocalFileSystem.getInstance().findFileByIoFile(Mockito.any())).thenReturn(new BinaryLightVirtualFile());
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: I think this line is longer than 80 chars

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+
+    Assert.assertEquals(new BinaryLightVirtualFile().toString(), FileUtils.virtualFileFromURI("file://foobar").toString());
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: I think this line is longer than 80 chars

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+import java.io.File;
+import java.net.URISyntaxException;
+
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:39]: Add a line between the different import packages and make it alphabetical. ie `com` then `java` then `org` then the` static imports` if any, same for them

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+    Assert.assertEquals(new BinaryLightVirtualFile().toString(), FileUtils.virtualFileFromURI("file://foobar").toString());
+  }
+
+  @PrepareForTest(LocalFileSystem.class)
+  @Test
+  public void testVirtualFileFromURINull() {
+    PowerMockito.mockStatic(LocalFileSystem.class);
+    PowerMockito.when(LocalFileSystem.getInstance()).thenThrow(URISyntaxException.class);
+    Assert.assertNull(FileUtils.virtualFileFromURI("foobar"));
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:84]: Add a blank line for consistency

**Review by EricHetti — CHANGES_REQUESTED** [review]
I think you have to check the line length as there are a few which are greater than 80 chars.. I didn't do a properly line length check but they seem too long

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 20]
```diff
@@ -0,0 +1,180 @@
+package org.wso2.lsp4intellij.utils;
```
Comment (Braavos6) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:16]: Copyright header is missing :)

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java (hunk 16 above)
Comment (ThomasPerkins1123) [reply src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java]: I have cut some of the larger ones but i'm not cutting every one that is above 80 as we did say were not going to be so fussy on formatting and the repository owner does not seem to care about line length as many of his are well above 100

**Review by EricHetti — APPROVED** [review]

**Review by JohnLBergqvist — APPROVED** [review]

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+
+    Assert.assertEquals(new BinaryLightVirtualFile().toString(),
+        FileUtils.virtualFileFromURI("file://foobar").toString());
+  }
+
+  @PrepareForTest(LocalFileSystem.class)
+  @Test
+  public void testVirtualFileFromURINull() {
+    PowerMockito.mockStatic(LocalFileSystem.class);
+    PowerMockito.when(LocalFileSystem.getInstance()).thenThrow(URISyntaxException.class);
```
Comment (petersear) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:82]: Why is this set up to throw an exception which doesn't get thrown?

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+
+    Assert.assertEquals(new BinaryLightVirtualFile().toString(),
+        FileUtils.URIToVFS("file://foobar").toString());
+  }
+
+  @PrepareForTest(LocalFileSystem.class)
+  @Test
+  public void testURIToVFSNull() {
+    PowerMockito.mockStatic(LocalFileSystem.class);
+    PowerMockito.when(LocalFileSystem.getInstance()).thenThrow(URISyntaxException.class);
```
Comment (petersear) [inline src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:137]: and this one?

**Comment by petersear** [comment]
Check the PR text.
`suggests` should be `suggested`
`i` should be `I`
The last sentence is incomplete.

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java (hunk 21 above)
Comment (ThomasPerkins1123) [reply src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:82]: this is done to reach a line in a try catch statement which only catches this specific exception.

File: src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java (hunk 22 above)
Comment (ThomasPerkins1123) [reply src/test/java/org/wso2/lsp4intellij/utils/FileUtilsTest.java:137]: same as above :
this is done to reach a line in a try catch statement which only catches this specific exception.

**Review by petersear — APPROVED** [review]

**Pull request merged**